kratos/utilities/parallel_distance_calculator.cpp

```diff
@@ -0,0 +1,630 @@
+    {
+        if (rGeometry[iii].Is(VISITED)) //identyfing the unknown node
+        {
+            const double distance = rGeometry[iii].FastGetSolutionStepValue(rDistanceVar);
+            avg_dist += distance;
+            for (unsigned int jjj = 0; jjj < TDim; jjj++)
+                d[jjj] += rDN_DX(iii, jjj) * distance;
+        }
+        else
+            unknown_node_index = iii;
+    }
+    avg_dist /= static_cast<double>(TDim);
+
+    //finalizing computation of discriminant
+    double c = -1.0;
+    double a = 0.0;
+    double b = 0.0;
+    for (unsigned int jjj = 0; jjj < TDim; jjj++)
+    {
+        a += rDN_DX(unknown_node_index, jjj) * rDN_DX(unknown_node_index, jjj);
+        b += d[jjj] * rDN_DX(unknown_node_index, jjj);
+        c += d[jjj] * d[jjj];
+    }
+    b *= 2.0;
+
+    //here we require (a*x^2 + b*x + c)^2 to be minimum (x represents the unknown distance)
+    //this implies setting to zero
+    //(a*x^2 + b*x + c)*(2ax+b) = 0
+    double distance;
+
+    double discriminant = b * b - 4.0 * a*c;
+
+    if (discriminant < 0.0) //here we solve (2ax+b) = 0
+    {
+//                  double numerator = 0.0;
+//                  double denominator = 0.0;
+//                  for(unsigned int i=0; i<TDim+1; i++)
+//                  {
+//                      for (unsigned int jjj = 0; jjj < TDim; jjj++)
+//                      {
+//                          if(i != unknown_node_index)
+//                            numerator += rDN_DX(unknown_node_index, jjj) * rDN_DX(i, jjj);
+//                          else
+//                            denominator += rDN_DX(unknown_node_index, jjj)*rDN_DX(unknown_node_index, jjj);
+//                      }
+//                  }
+//                  distance = - numerator/denominator;
+//
+//                  KRATOS_WATCH(rGeometry[unknown_node_index].Id());
+
+
+// 		KRATOS_WATCH(discriminant);
+        distance = -b / (2.0*a); //avg_dist ; //
+    }
+    else //in this case we solve (a*x^2 + b*x + c)=0
+    {
+        //(accurate) computation of the distance
+        //requires the solution of a*x^2+b*x+c=0
+        double q, root1, root2;
+        double sqrt_det = sqrt(discriminant);
+        if (a != 0.0)
+        {
+            if (b > 0) q = -0.5 * (b + sqrt_det);
+            else q = -0.5 * (b - sqrt_det);
+            root1 = q / a;
+            distance = root1;
+            if(std::abs(q) > 0.0) {
+                root2 = c / q;
+                if (root2 > root1) distance = root2;
+            }
+        }
+        else   //in this case we have a linear equation
+        {
+            distance = -c / b;
+        }
+    }
+
+    if(distance < 0.0)
+        distance = 1e-15;
+
+    rGeometry[unknown_node_index].SetLock();
+    rGeometry[unknown_node_index].FastGetSolutionStepValue(rDistanceVar) += distance*nodal_vol;
+    rGeometry[unknown_node_index].FastGetSolutionStepValue(rAreaVar) += nodal_vol;
+    rGeometry[unknown_node_index].UnSetLock();
```

Comment:
Wouldn't it better to get the node reference rather than always calling with node index?